AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -169,98 +161,89 @@ - (void)layout
+        [weakSelf setVideoPlaceholderImage:image];
+      }
+    });
+  }];
+}
+
+- (void)imageAtTime:(CMTime)imageTime completionHandler:(void(^)(UIImage *image))completionHandler
 {
   ASPerformBlockOnBackgroundThread(^{
     ASDN::MutexLocker l(_videoLock);
-    
-    AVAssetImageGenerator *imageGenerator = [[AVAssetImageGenerator alloc] initWithAsset:_asset];
-    imageGenerator.appliesPreferredTrackTransform = YES;
-    NSArray *times = @[[NSValue valueWithCMTime:CMTimeMake(0, 1)]];
-    
-    [imageGenerator generateCGImagesAsynchronouslyForTimes:times completionHandler:^(CMTime requestedTime, CGImageRef _Nullable image, CMTime actualTime, AVAssetImageGeneratorResult result, NSError * _Nullable error) {
-      
-      ASDN::MutexLocker l(_videoLock);
-      
-      // Unfortunately it's not possible to generate a preview image for an HTTP live stream asset, so we'll give up here
-      // http://stackoverflow.com/questions/32112205/m3u8-file-avassetimagegenerator-error
-      if (image && _placeholderImageNode.image == nil) {
-        [self setPlaceholderImage:[UIImage imageWithCGImage:image]];
-      }
-    }];
+
+    // Skip the asset image generation if we don't have any tracks available that are capable of supporting it
+    NSArray<AVAssetTrack *>* visualAssetArray = [_asset tracksWithMediaCharacteristic:AVMediaCharacteristicVisual];
+    if (visualAssetArray.count == 0) {
+      completionHandler(nil);
+      return;
+    }
+
+    AVAssetImageGenerator *previewImageGenerator = [AVAssetImageGenerator assetImageGeneratorWithAsset:_asset];
+    previewImageGenerator.appliesPreferredTrackTransform = YES;
+
+    [previewImageGenerator generateCGImagesAsynchronouslyForTimes:@[[NSValue valueWithCMTime:imageTime]]
+                                                completionHandler:^(CMTime requestedTime, CGImageRef image, CMTime actualTime, AVAssetImageGeneratorResult result, NSError *error) {
+                                                  if (error != nil && result != AVAssetImageGeneratorCancelled) {
+                                                    NSLog(@"Asset preview image generation failed with error: %@", error);
+                                                  }
+                                                  completionHandler(image ? [UIImage imageWithCGImage:image] : nil);
+                                                }];
   });
 }
 
-- (void)setPlaceholderImage:(UIImage *)image
+- (void)setVideoPlaceholderImage:(UIImage *)image
 {
   ASDN::MutexLocker l(_videoLock);
 
-  if (_placeholderImageNode == nil) {
+  if (_placeholderImageNode == nil && image != nil) {
     _placeholderImageNode = [[ASImageNode alloc] init];
     _placeholderImageNode.layerBacked = YES;
     _placeholderImageNode.contentMode = ASContentModeFromVideoGravity(_gravity);
   }
 
   _placeholderImageNode.image = image;
 
-  dispatch_async(dispatch_get_main_queue(), ^{
+  ASPerformBlockOnMainThread(^{
```

Comment:
I really love the attention to detail in all of your patches.  Every other line I be like, YUSSSS!  haha